File: serene/src/Serene.Web/Modules/Administration/Role/RoleRow.cs
namespace Serene.Administration;

[ConnectionKey("Default"), Module("Administration"), TableName("Roles")]
[DisplayName("Roles"), InstanceName("Role")]
[ReadPermission(PermissionKeys.Security)]
[ModifyPermission(PermissionKeys.Security)]
[LookupScript]
public sealed class RoleRow : Row<RoleRow.RowFields>, IIdRow, INameRow
{
    [DisplayName("Role Id"), Identity, ForeignKey("Roles", "RoleId"), LeftJoin("jRole"), IdProperty]
    public int? RoleId { get => fields.RoleId[this]; set => fields.RoleId[this] = value; }

    [DisplayName("Role Name"), Size(100), NotNull, QuickSearch, NameProperty]
    public string RoleName { get => fields.RoleName[this]; set => fields.RoleName[this] = value; }

    public class RowFields : RowFieldsBase
    {
        public Int32Field RoleId;
        public StringField RoleName;
    }
}

File: serene/src/Serene.Web/Modules/ServerTypes/Administration/RoleRow.ts
import { getLookup, getLookupAsync, fieldsProxy } from "@serenity-is/corelib";

export interface RoleRow {
    RoleId?: number;
    RoleName?: string;
}

export abstract class RoleRow {
    static readonly idProperty = 'RoleId';
    static readonly nameProperty = 'RoleName';
    static readonly localTextPrefix = 'Administration.Role';
    static readonly lookupKey = 'Administration.Role';

    /** @deprecated use getLookupAsync instead */
    static getLookup() { return getLookup<RoleRow>('Administration.Role') }
    static async getLookupAsync() { return getLookupAsync<RoleRow>('Administration.Role') }

    static readonly deletePermission = 'Administration:Security';
    static readonly insertPermission = 'Administration:Security';
    static readonly readPermission = 'Administration:Security';
    static readonly updatePermission = 'Administration:Security';

    static readonly Fields = fieldsProxy<RoleRow>();
}